File: truss-compiler/src/main/java/truss/compiler/MessageCollection.java
package truss.compiler;

import org.apache.commons.lang.Validate;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MessageCollection {
    private final List<Message> messages = new ArrayList<>();
    private final List<Message> unmodifiableMessages = Collections.unmodifiableList(messages);
    private int errors;
    private int warnings;
    private int infos;

    public boolean hasErrors() {
        return errors > 0;
    }

    public int getErrors() {
        return errors;
    }

    public boolean hasErrorsOrWarnings() {
        return errors > 0 || warnings > 0;
    }

    public int getWarnings() {
        return warnings;
    }

    public boolean hasMessages() {
        return messages.size() > 0;
    }

    public int getInfos() {
        return infos;
    }

    public List<Message> getMessages() {
        return unmodifiableMessages;
    }

    public void add(Message message) {
        Validate.notNull(message, "message");

        messages.add(message);

        switch (message.getType().getSeverity()) {
            case ERROR: errors++; break;
            case WARN: warnings++; break;
            case INFO: infos++; break;
        }
    }

    public void add(MessageType type, Object... args) {
        add(new Message(type, args));
    }

    public void add(MessageType type, String fileName, Object... args) {
        add(new Message(type, fileName, args));
    }

    public void add(MessageType type, Span span, Object... args) {
        add(new Message(type, span, args));
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        for (Message message : messages) {
            sb.append(message.toString());
            sb.append('\n');
        }

        return sb.toString();
    }
}

File: Truss.Compiler/MessageCollection.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;

namespace Truss.Compiler {
    public class MessageCollection {
        private readonly List<Message> _messages= new List<Message>();

        public MessageCollection() {
            Messages = new ReadOnlyCollection<Message>(_messages);
        }

        public bool HasErrors {
            get { return Errors > 0; }
        }

        public int Errors { get; private set; }

        public bool HasErrorsOrWarnings {
            get { return Errors > 0 || Warnings > 0; }
        }

        public int Warnings { get; private set; }

        public bool HasMessages {
            get { return _messages.Count > 0; }
        }

        public int Infos { get; private set; }

        public IList<Message> Messages { get; private set; }

        public void Add(Message message) {
            if (message == null) {
                throw new ArgumentNullException("message");
            }

            _messages.Add(message);

            switch (message.Type.Severity) {
                case Severity.Error:
                    Errors++;
                    break;
                case Severity.Warn:
                    Warnings++;
                    break;
                case Severity.Info:
                    Infos++;
                    break;
            }
        }

        public void Add(MessageType type, params object[] args) {
            Add(new Message(type, args));
        }

        public void Add(MessageType type, string fileName, params object[] args) {
            Add(new Message(type, fileName, args));
        }

        public void Add(MessageType type, Span span, params object[] args) {
            Add(new Message(type, span, args));
        }

        public override string ToString() {
            var sb = new StringBuilder();

            foreach (var message in _messages) {
                sb.Append(message);
                sb.Append(Environment.NewLine);
            }

            return sb.ToString();
        }
    }
}
